op-batcher: syncActions is aware of safe=unsafe edge case

geoknee opened the pull request:
Follow up to #13287 with some more tests and edge case awareness. 

Also avoiding use of "oldest unsafe block" in cases where there may not be any unsafe blocks. Preferring "next safe block" instead.

geoknee committed:
add additional test case for computeSyncActions

geoknee committed:
fixup test

geoknee committed:
check for unsafe = safe edge case & replace oldestUnsafeBlock with nextSafeBlock

geoknee committed:
add test case for no progress and safe=unsafe

geoknee committed:
refine log

geoknee committed:
rename variable

geoknee committed:
harmonize log ordering and labels

sebastianst commented on op-batcher/batcher/sync_actions.go:
```diff
@@ -133,7 +140,7 @@ func computeSyncActions[T channelStatuser](newSyncStatus eth.SyncStatus, prevCur
 
 	start := newestBlockInStateNum + 1
-	end := youngestUnsafeBlockNum
+	end := newSyncStatus.UnsafeL2.Number
```
How is the case handled that there simply aren't new unsafe blocks, so safe = unsafe?

geoknee replied:
Fixed here [b689833](https://github.com/ethereum-optimism/optimism/pull/13292/commits/b68983330d089ec77e8af60f12d88a6388192047)

sebastianst commented on op-batcher/batcher/sync_actions_test.go:
```diff
@@ -226,6 +243,18 @@ func TestBatchSubmitter_computeSyncActions(t *testing.T) {
 				blocksToLoad:    &inclusiveBlockRange{105, 109},
 			},
 		},
+		{name: "no progress + unsafe=safe",
```
we need a test case with safe=unsafe where there are blocks in state

geoknee replied:
Added here [b689833](https://github.com/ethereum-optimism/optimism/pull/13292/commits/b68983330d089ec77e8af60f12d88a6388192047)

geoknee committed:
tighten up test behaviour for expectedLogs

geoknee committed:
add test case: no progress + unsafe=safe + blocks in state

and fix behaviour so we don't try to fetch unsafe blocks if there aren't any, even when there are blocks in state

geoknee committed:
typo
